### DISCUSSION TITLE
How to inherit from Trackball class

### DISCUSSION THREAD

**Discussion by smacl6301** [vsg-dev/VulkanSceneGraph#1226] [General]
I'm looking to write a modified Trackball class to change some default functionality and add some new functionality of my own.  However, when I create a a new class derived from Trackball, none of the virtual functions from Trackball are overridden and the debugger doesn't even let me set breakpoints on them.  As someone new to vsg coding, I guess I'm missing something very obvious here.  A simplified version of my class code is shown below

```
namespace vsg
{
	class MyTrackball : public Trackball
	{
	public:
		void rotate(double angle, const dvec3& axis)
		{
			dvec3 axisb = axis;
			axisb.z = 0;
			vsg::Trackball::rotate(angle, axisb);
		}
	};
};
```
In my main application the code to instantiate the trackball copied from the example code is;

`	viewer->addEventHandler(vsg::MyTrackball::create(camera));`

**Comment by vsg-dev** [comment]
To make sure you are overriding the methods with the correct parameters use the override keyword after the method declaration, this will let the compiler you are intending to override a parent classes virtual method, and will emit an compile error if you've got something mismatched.

**Comment by smacl6301** [comment]
It looks like the issue is that `vsg::MyTrackball::create(camera)` calls the base class constructor via `Inherit<vsg::Visitor,vsg::Trackball>`.  Call stack is below

```
 	SCCW64R.exe!vsg::Trackball::Trackball(vsg::ref_ptr<vsg::Camera> camera, vsg::ref_ptr<vsg::EllipsoidModel> ellipsoidModel) Line 27	C++
 	SCCW64R.exe!vsg::Inherit<vsg::Visitor,vsg::Trackball>::create<vsg::ref_ptr<vsg::Camera> &>(vsg::ref_ptr<vsg::Camera> & <args_0>) Line 37	C++

```
Looking at Inherit, we can't use create here as we can't inherit from `Inherit ` in MyTrackball due to multiple inheritance ambiguities and thus can't use `create` here.  My attempted solution was

```
	vsg::ref_ptr<vsg::MyTrackball> trackball(camera);
	viewer->addEventHandler(trackball);

```

But that caused issues with const conflicts

```
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440: 'initializing': cannot convert from 'T *const ' to 'T *'
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:         with
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:         [
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:             T=vsg::Camera
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:         ]
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:         and
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:         [
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:             T=vsg::Visitor
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2440:         ]
1>C:\cpp\VulkanSceneGraph\include\vsg\core\ref_ptr.h(45,17): error C2439: 'vsg::ref_ptr<vsg::Visitor>::_ptr': member could not be initialized

```
I've tried a few combinations to get rid of the const problem here without much luck, not sure how to proceed here other than copying the full source of trackball into my project and working on that.

**Comment by robertosfield** [comment]
I think you are conflating different bits of functionality rather than keeping things separate.  

The vsg::Inherit template class is used as a parent class of the new subclass to provide common RTTI and create functionality that does not exist in C++ itself but is useful for general C++ programming and scene graphs in particular.  The template approach is an example of the "Curiously Recurring Template Pattern" or CRTP for short, a C++ idiom that helps solve the problem of implement consistent API implementations in subclasses.

It isn't essential to subclass from vsg::Inherit<> when subclassing from VSG classes like vsg::Visitor but for objects that RTTI or ensuring objects are created on the heap with the appropriate ref_ptr<> returned to avoid possible memory leaks it's far more helpful in implementing base class methods for you.  

The Inherit::create() template method in particular provides a safer and easier to use replacement for new.  With new it's possible to assign it to a C pointer and not have the compiler complain, then the developer could start doing unsafe things like deleting an object that is shared with other objects causing dangling pointers, or using it to access a dangling pointer that still points to a object that has already been deleted.  Passing back ref_ptr<> makes sure memory manager is safe right from the start of an objects life, even if an exception is thrown everything is still cleaned up correctly.

From your description you are mixing up the inheritance,  You can implement your own trackball from scratch and so subclass from vsg::Visitor directly as vsg::Trackball does, then implement the apply(..) methods relevant to event handling.  This approach for handling events leverages the VSG''s support for the "Visitor Design Pattern."

If you want to subclass from vsg::Trackball then simple do:

~~~ cpp
class MyTrackball: public vsg::Inherit<vsg::Trackball, MyTrackball>
{
...
};
~~~

In the vsgTutorial online book that I've started writing I cover some of these topics so worth reading:

https://vsg-dev.github.io/vsgTutorial/

**Reply by smacl6301** [reply]
Hi Robert,

Thanks for the feedback.  Using

```
class MyTrackball: public vsg::Inherit<vsg::Trackball, MyTrackball>
{
};

.
.
viewer->addEventHandler(MyTrackball::create(camera));

```
results in a different compilation error.  

`C:\cpp\VulkanSceneGraph\include\vsg\core\Inherit.h(37,50): error C2665: 'MyTrackball::MyTrackball': no overloaded function could convert all the argument types`

I think the simplest approach will be to implement my own trackball class from scratch based on the existing trackball code.  I'd assumed that as the Trackball class included virtual functions the intention was that other classes could be derived from it but think i'm mistaken here.

**Comment by robertosfield** [comment]
Without seeing all your code I can't say what you are doing wrong.  Rather than try to guess I quickly threw together a modification of vsgviewer example program in vsgExamples to illustrate how you subclass from Trackball.  I have put this in a CiustomTracball branch:

https://github.com/vsg-dev/vsgExamples/compare/master...CustomTrackball

I don't plan to merge this example as it's really just thrown together with no plan to illustrate what the code looks like.

**Reply by smacl6301** [reply]
Hi Robert, many thanks for the example.  I was getting the use of Inherit in the ctor wrong in my attempt.

**Comment by hytel** [comment]
I've gotten bitten by this issue just now.  Would love to see the CustomTrackball example, but it isn't accessible anymore, is it?

**Reply by robertosfield** [reply]
I haven't had time to review and merge the vsgturntable example yet but it's there in the vsgExamples PR:

https://github.com/vsg-dev/vsgExamples/pull/354

**Comment by hytel** [comment]
Drat it.  The problem is when you get something wrong with templates, there are like 10 pages in the error output of Visual Studio for just one issue and it's hard to track down what cause is.  I modeled my inheritance after your example but it threw an error - causing me to think the class was still defined wrong.  It wasn't, it's in the CPP constructor instantiation that was the problem.  I have finally found it and fixed it.  My situation was a bit more complex because I am trying to derive from a templated class that in-turn derives from the vsg::Inherit base.  But works now!
